### PR TITLE
[SER-281] Log Errors

### PR DESCRIPTION
**Pull request by mountHouli**
### Jira Ticket(s)
[SER-281](https://jira.amida.com/browse/SER-281)

#### Test 1

`yarn test`.

#### Test 2 (optional): 

Step A: Setup logging

In `format.test.js`, in each of the `beforeEach()` callbacks that specify a jest mock function (e.g. `spy = jest.fn()`), add this logging statement: `console.log(expectedMessage);`

Step B: Setup script

In `package.json`, add this script `"test-with-logging": "yarn install && jest --verbose=false",`

Step C: Run it

`yarn test-with-logging`

Results: Logged to your terminal should be each of the desired outcomes listed in the jira ticket.

### CODE REVIEW

**Review by orndorffgrant — APPROVED** [review]
Very 🔥

File: src/index.js [hunk 1]
```diff
@@ -57,12 +47,17 @@ const jsonFormat = printf((info) => {
...
+        }
+    };
+
+    if (info.stack) {
+        logObject.err = {
             name: info.name,
             stack: info.stack
-        }
-    });
+        };
```
Comment (rmharrison) [inline src/index.js:100]: **Open question**
- `code: info.code`?
- Conditional logic to only include `info.code` if it belonged to an error?
  * Probably difficult since everything is flattened. :'(

File: src/index.js [hunk 2]
```diff
@@ -97,9 +92,28 @@ function configuredFormatter(options = {}) {
     throw new Error(`${typeFormat} is not json or console.`);
 }
 
+function parseInfo(infoObj) {
+    return _.omit(infoObj, [
```
Comment (rmharrison) [inline src/index.js:163]: Add code, if including in `err`

**Review by rmharrison — CHANGES_REQUESTED** [review]
## TL;DR
The function signatures don't work quite how I expected. Some of this is bunyan vs winston signature standards.

Most are edge-cases that behave unexpectedly.

- Which are behaving as you expect? 
- Which, if any, *should* be supported?

And thank you for explicitly enumerating all of the test cases!

In 20+ test cases, the only enumerated test case that doesn't behave as I expected is (B), the concatenation of LoggerMsg and ErrorMsg with `logger.error("LoggerMsg", new Error("ErrorMsg"))`

FFR: To reproduce, copy/paste the lines from the [README.md:L13-32](https://github.com/amida-tech/winston-json-formatter/blame/4e4c7c9bc345d098758f3e70cc3a3e004cacd59a/README.md#L13-L32) and use the `logger.error` signatures enumerated below.

## (A) Resolved: I was surprised that the `err` first signature didn't work (this is a difference with bunyan)
```
logger.error(new Error('abc'), 'msg')
{
  "service": "test-service",
  "logger": "Winston-JSON-Formatter",
  "hostname": "...",
  "level": "error",
  "msg": {},
  "meta": {
    "service": {
      "version": "1.0.0",
      "node_env": ""
    },
    "logger": {
      "time": "2019-04-09T23:00:31.789Z"
    },
    "event": {}
  }
}
```
- No `err`, but this is expected from the winston standard.

Thank you for linking the specific winston fix PR in your JIRA issue: https://github.com/winstonjs/winston/pull/1562#issue-242542523
Here, they enumerate the supported signatures, which *does not* include `err` first.

## (B) Unresolved: LoggerMsg and ErrorMsg behaviour
Unexpected...
```
> logger.error('LoggerMsg', new Error('ErrorMsg'))
{
  "service": "test-service",
  "logger": "Winston-JSON-Formatter",
  "hostname": "...",
  "level": "error",
  "msg": "LoggerMsgErrorMsg", # <~~~ Concatated LoggerMsg and ErrorMsg
  "meta": {
    "service": {
      "version": "1.0.0",
      "node_env": ""
    },
    "logger": {
      "time": "2019-04-09T23:04:01.682Z"
    },
    "event": {}
  },
  "err": {
    "stack": "..." # <~~~ err.name doesn't appear
  }
}
```
- LoggerMsg and ErrorMsg are concatinated
- `err.name` doesn't appear

Expected...
```
> logger.error(new Error('ErrorMsg'))
{
  "service": "test-service",
  "logger": "Winston-JSON-Formatter",
  "hostname": "...",
  "level": "error",
  "msg": "ErrorMsg",
  "meta": {
    "service": {
      "version": "1.0.0",
      "node_env": ""
    },
    "logger": {
      "time": "2019-04-09T23:05:30.698Z"
    },
    "event": {}
  },
  "err": {
    "name": "Error",
    "stack": "..."
  }
}
```

`err.name` appears as expected.

## (C) Unresolved: Unsupported signature: Err, {}
```
> logger.error(new Error('ErrorMsg'), {a: 'a', b: 'b'})
{
  "service": "test-service",
  "logger": "Winston-JSON-Formatter",
  "hostname": "...",
  "level": "error",
  "msg": {},
  "meta": {
    "service": {
      "version": "1.0.0",
      "node_env": ""
    },
    "logger": {
      "time": "2019-04-09T23:11:57.836Z"
    },
    "event": {
      "a": "a",
      "b": "b"
    }
  }
}
```
- No `err`

## (D) Unresolved: Unsupported signature: FunctionMsg, Err, {}
```
> logger.error('LoggerMsg', new Error('ErrorMsg'), {a: 'a', b: 'b'})
{
  "service": "test-service",
  "logger": "Winston-JSON-Formatter",
  "hostname": "Hargrove",
  "level": "error",
  "msg": "LoggerMsgErrorMsg",
  "meta": {
    "service": {
      "version": "1.0.0",
      "node_env": ""
    },
    "logger": {
      "time": "2019-04-09T23:12:57.049Z"
    },
    "event": {} # <~~~ {a: 'a', b: 'b'} don't appear here
  },
  "err": {
    "stack": "..."
  }
}

```

## Clarification
Currently, `code` is not "magic" and appears under `meta.event`
If `code` is part of an Error object, should it appear under `err.code`?

**Comment by rmharrison** [comment]
Please update readme:
- Update JSON log format
- New screenshot for console logger

File: src/index.js (hunk 1 above)
Comment (mountHouli) [reply src/index.js:100]: Indeed. This cannot be done.

File: src/index.js (hunk 2 above)
Comment (mountHouli) [reply src/index.js:163]: We want `code` to _not_ get filtered out by `parseInfo()`. This is partially because
1. There is no way to determine if a member named `code` is a member of an error or of a different type of object passed into the logger.
2. If we filter out `code`, the logical extension of that would be to filter out any/every possible non-standard member of an object of type `Error`, which is impossible and also makes this library dependent on our error handling philosophy.
3. We do want `code` (and any other custom properties of objects of type `Error`) to show up in `meta.event`.

**Comment by mountHouli** [comment]
## Regarding item (B)

The concatenated messages are a winston bug. See https://github.com/winstonjs/winston/issues/1634

The lack of `err.name` / `info.name` is a winston bug. See https://github.com/winstonjs/winston/issues/1635

## Regarding item (C)

Fixed. See commit a8ed890b60d6583bc87cba424417fc29bd81c4c2

## Regarding item (D)

This case is not supported by winston. See examples in https://github.com/winstonjs/winston/pull/1562

**Review by rmharrison — APPROVED** [review]
